File: backend/app/Services/DashboardService.php
<?php

namespace App\Services;

use App\Models\Order;
use App\Models\ProductionOrder;
use App\Models\QualityControl;
use App\Models\Stock;
use Illuminate\Support\Facades\DB;

class DashboardService
{
    public function getStats(): array
    {
        return [
            'total_orders' => Order::count(),
            'active_orders' => Order::whereIn('status', ['Pending', 'In Production', 'Quality Control'])->count(),
            'delayed_orders' => Order::where('status', 'Delayed')->count(),
            'completed_orders' => Order::where('status', 'Completed')->count(),
            'critical_stocks' => Stock::whereRaw('quantity_meter < critical_level')->count(),
            'running_productions' => ProductionOrder::where('status', 'Running')->count(),
            'defect_rate' => $this->getDefectRate(),
            'production_efficiency' => $this->getProductionEfficiency(),
        ];
    }

    public function getWeeklyProduction(): array
    {
        $data = DB::table('production_orders')
            ->select(DB::raw('DATE(created_at) as date, COUNT(*) as count'))
            ->where('created_at', '>=', now()->subDays(7))
            ->groupBy(DB::raw('DATE(created_at)'))
            ->orderBy('date')
            ->get();

        return $data->map(fn($row) => [
            'date' => $row->date,
            'count' => $row->count,
        ])->toArray();
    }

    public function getMonthlyEfficiency(): array
    {
        $data = DB::table('production_orders')
            ->select(DB::raw('DATE_FORMAT(created_at, "%Y-%m") as month, AVG(progress_percentage) as efficiency'))
            ->where('created_at', '>=', now()->subMonths(6))
            ->groupBy(DB::raw('DATE_FORMAT(created_at, "%Y-%m")'))
            ->orderBy('month')
            ->get();

        return $data->map(fn($row) => [
            'month' => $row->month,
            'efficiency' => round($row->efficiency, 1),
        ])->toArray();
    }

    public function getDefectDistribution(): array
    {
        return DB::table('quality_controls')
            ->select('defect_type', DB::raw('SUM(defect_quantity) as value'))
            ->where('defect_type', '!=', 'none')
            ->groupBy('defect_type')
            ->get()
            ->map(fn($row) => [
                'name' => $row->defect_type,
                'value' => (int) $row->value,
            ])
            ->toArray();
    }

    private function getDefectRate(): float
    {
        $total = DB::table('quality_controls')
            ->sum(DB::raw('defect_quantity + passed_quantity'));
        $defects = DB::table('quality_controls')->sum('defect_quantity');
        return $total > 0 ? round(($defects / $total) * 100, 2) : 0;
    }

    private function getProductionEfficiency(): float
    {
        return round(ProductionOrder::avg('progress_percentage') ?? 0, 1);
    }
}

File: backend-dotnet/Services/DashboardService.cs
using Microsoft.EntityFrameworkCore;
using TexTrack.Api.Data;

namespace TexTrack.Api.Services;

public class DashboardService
{
    private readonly AppDbContext _db;
    public DashboardService(AppDbContext db) => _db = db;

    private static readonly string[] ActiveStatuses = { "Pending", "In Production", "Quality Control" };

    public async Task<object> GetStatsAsync()
    {
        return new
        {
            total_orders = await _db.Orders.CountAsync(),
            active_orders = await _db.Orders.CountAsync(o => ActiveStatuses.Contains(o.Status)),
            delayed_orders = await _db.Orders.CountAsync(o => o.Status == "Delayed"),
            completed_orders = await _db.Orders.CountAsync(o => o.Status == "Completed"),
            critical_stocks = await _db.Stocks.CountAsync(s => s.QuantityMeter < s.CriticalLevel),
            running_productions = await _db.ProductionOrders.CountAsync(p => p.Status == "Running"),
            defect_rate = await GetDefectRateAsync(),
            production_efficiency = await GetProductionEfficiencyAsync(),
        };
    }

    public async Task<object> GetWeeklyProductionAsync()
    {
        var since = DateTime.UtcNow.AddDays(-7);
        var rows = await _db.ProductionOrders
            .Where(p => p.CreatedAt >= since)
            .Select(p => p.CreatedAt)
            .ToListAsync();

        return rows
            .GroupBy(d => d.ToString("yyyy-MM-dd"))
            .OrderBy(g => g.Key)
            .Select(g => new { date = g.Key, count = g.Count() })
            .ToList();
    }

    public async Task<object> GetMonthlyEfficiencyAsync()
    {
        var since = DateTime.UtcNow.AddMonths(-6);
        var rows = await _db.ProductionOrders
            .Where(p => p.CreatedAt >= since)
            .Select(p => new { p.CreatedAt, p.ProgressPercentage })
            .ToListAsync();

        return rows
            .GroupBy(r => r.CreatedAt.ToString("yyyy-MM"))
            .OrderBy(g => g.Key)
            .Select(g => new { month = g.Key, efficiency = Math.Round(g.Average(x => x.ProgressPercentage), 1) })
            .ToList();
    }

    public async Task<object> GetDefectDistributionAsync()
    {
        var rows = await _db.QualityControls
            .Where(q => q.DefectType != "none")
            .GroupBy(q => q.DefectType)
            .Select(g => new { name = g.Key, value = g.Sum(x => x.DefectQuantity) })
            .ToListAsync();
        return rows;
    }

    private async Task<double> GetDefectRateAsync()
    {
        int total = await _db.QualityControls.SumAsync(q => (int?)(q.DefectQuantity + q.PassedQuantity)) ?? 0;
        int defects = await _db.QualityControls.SumAsync(q => (int?)q.DefectQuantity) ?? 0;
        return total > 0 ? Math.Round((double)defects / total * 100, 2) : 0;
    }

    private async Task<double> GetProductionEfficiencyAsync()
    {
        var avg = await _db.ProductionOrders.AverageAsync(p => (double?)p.ProgressPercentage) ?? 0;
        return Math.Round(avg, 1);
    }
}
